xdsl/dialects/pdl_interp.py

```diff
@@ -0,0 +1,664 @@
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLInterpOps/#pdl_interpget_value_type-pdl_interpgetvaluetypeop
+    """
+
+    name = "pdl_interp.get_value_type"
+    T: ClassVar = VarConstraint("T", base(TypeType) | base(RangeType[TypeType]))
+    value = operand_def(ValueConstrFromResultConstr(T))
+    result = result_def(T)
+
+    assembly_format = "`of` $value `:` type($result) attr-dict"
+
+    def __init__(self, value: SSAValue) -> None:
+        super().__init__(
+            operands=[value],
+            result_types=[
+                RangeType(TypeType())
+                if isinstance(value.type, RangeType)
+                else TypeType()
+            ],
+        )
+
+
+@irdl_op_definition
+class ReplaceOp(IRDLOperation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLInterpOps/#pdl_interpreplace-pdl_interpreplaceop
+    """
+
+    name = "pdl_interp.replace"
+    input_op = operand_def(OperationType)
+    repl_values = var_operand_def(ValueType | RangeType[ValueType])
+
+    assembly_format = (
+        "$input_op `with` ` ` `(` ($repl_values^ `:` type($repl_values))? `)` attr-dict"
+    )
+
+    def __init__(self, input_op: SSAValue, repl_values: list[SSAValue]) -> None:
+        super().__init__(operands=[input_op, repl_values])
+
+
+@irdl_op_definition
+class CreateAttributeOp(IRDLOperation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLInterpOps/#pdl_interpcreate_attribute-pdl_interpcreateattributeop
+    """
+
+    name = "pdl_interp.create_attribute"
+    value = prop_def(AnyAttr())
+    attribute = result_def(AttributeType)
+
+    assembly_format = "$value attr-dict-with-keyword"
+
+    def __init__(self, value: Attribute) -> None:
+        super().__init__(properties={"value": value}, result_types=[AttributeType()])
+
+
+@irdl_op_definition
+class CreateOperationOp(IRDLOperation):
+    """
+    https://mlir.llvm.org/docs/Dialects/PDLInterpOps/#pdl_interpcreate_operation-pdl_interpcreateoperationop
+    """
+
+    name = "pdl_interp.create_operation"
+    constraint_name = prop_def(StringAttr, prop_name="name")
+    input_attribute_names = prop_def(ArrayAttr, prop_name="inputAttributeNames")
+    inferred_result_types = opt_prop_def(UnitAttr, prop_name="inferredResultTypes")
+
+    input_operands = var_operand_def(ValueType | RangeType[ValueType])
+    input_attributes = var_operand_def(AttributeType | RangeType[AttributeType])
+    input_result_types = var_operand_def(TypeType | RangeType[TypeType])
+
+    result_op = result_def(OperationType)
+
+    irdl_options = [AttrSizedOperandSegments(as_property=True)]
+
+    # assembly_format = (
+    #     "$name (`(` $input_operands^ `:` type($input_operands) `)`)?"
+    #     "` `custom<CreateOperationOpAttributes>($inputAttributes, $inputAttributeNames)"
+    #     "custom<CreateOperationOpResults>($inputResultTypes, type($inputResultTypes), $inferredResultTypes)"
+    #     "attr-dict"
+    # )
+    # # unexpected token 'custom'
```

Comment:
Can you please add a bit more of a comment here? In general we don't want commented code, but if the broken assembly format is a useful thing to track that's OK, it would be good to add this in text here.